content/en/altinity-kb-setup-and-maintenance/altinity-kb-clean-replication-ddl-queue.md

```diff
@@ -0,0 +1,205 @@
+description: >
+    This article describes how to detect possible problems in the `replication_queue` and `distributed_ddl_queue` and how to clean if needed.
+---
+
+# How to check the replication problems:
+
+1. check `system.replicas` first, cluster-wide. It allows to check if the problem is local to some replica or global, and allows to see the exception.
+   allows to answer the following questions:
+   - Are there any ReadOnly replicas?
+   - Is there the connection to zookeeper active?
+   - Is there the exception during table init? (`Code: 999. Coordination::Exception: Transaction failed (No node): Op #1`)
+  
+2. Check `system.replication_queue`. How many tasks there / are they moving / is there some very old tasks there?:
+   - You can use this qkb article query: https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-replication-queue/
+   - Check if there are tasks with a high number of `num_tries` or `num_postponed` and `postponed_reason` this is a sign of stuck tasks.
+   - Check the problematic parts affecting the stuck tasks. You can use columns `new_part_name` or `parts_to_merge`
+   - Check which type is the task. If it is `MUTATE_PART` then it is a mutation task. If it is `MERGE_PARTS` then it is a merge task. These tasks can be deleted from the replication queue but `GET_PARTS` should not be deleted.
+
+3. Check system.errors
+
+4. Check system.mutations
+   - You can check that in the replication queue are stuck tasks of type `MUTATE_PART`, and that those mutations are still executing `system.mutations` using column `is_done`
+
+5. Find the moment when problem started and collect / analyze / preserve logs from that moment. It is usually during the first steps of a restart/crash
+
+6. Use `part_log` and `system.parts` to gather information of the parts related with the stuck tasks in the replication queue.
+   - Check if those parts exist and are active from `system.parts`.
+   - Extract the part history from `system.part_log`
+
+7. Example query from `part_log`
+
+```sql
+SELECT hostName(), * FROM 
+cluster('all-sharded',system.part_log)
+WHERE
+    hostName() IN ('chi-live-live-2-0-0','chi-live-live-2-2-0','chi-live-live-2-1-0')
+    AND table = 'session_events_local'
+    AND database = 'insights'
+    AND part_name in ('20230411_33631_33654_3')
+```
+
+9. If there are no errors, just everything get slower - check the load (usual system metrics)
+
+
+## Common problems & solutions
+
+
+- If the replication queue does not have any Exceptions only postponed reasons without exceptions just leave ClickHouse do Merges/Mutations and it will eventually catch up and reduce the number of tasks in `replication_queue`. If Delay in queue is going up actions may be needed.
+
+
+- Unknown stuff
+  - try to SYSTEM RESTART REPLICA
+  - try to restart ClickHouse
+
+## don't run things below blindly!!! they are dangerous! you can make it worse! Only if you clearly understand the problem and what that solution will do.
```

Comment:
I'm a bit afraid of publishing that because of that... It can be harmful, and fixing the ClickHouse after unsuccessful attempts of fixing that is more cumbersome, and data can be damaged.

Reply:
Yes agree i can delete it and rephrase. This also is a WIP, need to add a section for cleaning the distributed_ddl_queue that I will probably need to write it with care. I'll fix many stuff in next commit